internal/controller/services/monitoring/monitoring_controller.go

```diff
@@ -98,3 +115,140 @@ func (h *serviceHandler) NewReconciler(ctx context.Context, mgr ctrl.Manager) er
 
 	return nil
 }
+
+func deployTempo(ctx context.Context, rr *types.ReconciliationRequest, tracingSpec *serviceApi.TracesSpec, namespace string) error {
+	if tracingSpec == nil {
+		// Ensure Tempo instance is absent
+		return removeTempoInstance(ctx, rr, namespace)
+	}
+
+	// local copy of tracingSpec to avoid mutating the live Spec
+	localTracingSpec := *tracingSpec
+
+	// Validate the backend value against a whitelist
+	allowedBackends := map[string]bool{
+		"pv":  true,
+		"s3":  true,
+		"gcs": true,
+	}
+
+	if localTracingSpec.Storage.Backend == "" {
+		localTracingSpec.Storage.Backend = "pv"
+	}
+
+	if !allowedBackends[localTracingSpec.Storage.Backend] {
+		return fmt.Errorf("unknown backend: %s", localTracingSpec.Storage.Backend)
+	}
+
+	if localTracingSpec.Storage.Backend == "pv" {
+		return deployTempoMonolithic(rr, &localTracingSpec, namespace)
+	}
+	return deployTempoStack(rr, &localTracingSpec, namespace)
+}
+
+func deployTempoMonolithic(rr *types.ReconciliationRequest, tracingSpec *serviceApi.TracesSpec, namespace string) error {
+	tempo := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "tempo.grafana.com/v1alpha1",
+			"kind":       "TempoMonolithic",
+			"metadata": map[string]interface{}{
+				"name":      "tempo",
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"multitenancy": map[string]interface{}{
+					"enabled": true, // Required for OpenShift
+				},
+				"storage": map[string]interface{}{
+					"traces": map[string]interface{}{
+						"backend": "pv",
+						"size":    tracingSpec.Storage.Size, // Optional
+					},
+				},
+			},
+		},
+	}
+
+	if err := rr.AddResources(tempo); err != nil {
+		return errors.New("failed to deploy TempoMonolithic")
+	}
+	return nil
+}
+
+func deployTempoStack(rr *types.ReconciliationRequest, tracingSpec *serviceApi.TracesSpec, namespace string) error {
+	// Validate required fields for TempoStack
+	if tracingSpec.Storage.Secret == "" {
```

Comment:
you can probably move such validation to the CRD usin a cel expression, i.e. https://github.com/opendatahub-io/opendatahub-operator/blob/kueue-migration/api/infrastructure/v1alpha1/hardwareprofile_types.go#L72